src/functionals.jl

```diff
@@ -22,13 +22,32 @@ function expectation(distr::UnivariateDistribution, g::Function)
     expectation(distr, g, 1e-10)
 end
 
+function expectation(distr::MultivariateDistribution, g::Function; nsamples::Int=100, rng::AbstractRNG=GLOBAL_RNG)
+    nsamples > 0 || throw(ArgumentError("number of samples should be > 0"))
+    # We use a function barrier to work around type instability of `sampler(dist)`
+    return mcexpectation(rng, g, sampler(distr), nsamples)
+end
+
+mcexpectation(rng, f, sampler, n) = sum(f, rand(rng, sampler) for _ in 1:n) / n
+
 ## Leave undefined until we've implemented a numerical integration procedure
 # function entropy(distr::UnivariateDistribution)
 #     pf = typeof(distr)<:ContinuousDistribution ? pdf : pmf
 #     f = x -> pf(distr, x)
 #     expectation(distr, x -> -log(f(x)))
 # end
 
-function kldivergence(P::UnivariateDistribution, Q::UnivariateDistribution)
-    expectation(P, x -> let p = pdf(P,x); (p > 0)*log(p/pdf(Q,x)) end)
+function safe_logdiff(P, Q)
+    return function _safe_logdiff(x)
+        logp = logpdf(P, x)
+        return (logp > oftype(logp, -Inf)) * (logp - logpdf(Q, x))
+    end
+end
+
+function kldivergence(P::Distribution{V}, Q::Distribution{V}; kwargs...) where {V<:VariateForm}
+    expectation(P, safe_logdiff(P, Q))
+end
+
+function kldivergence(P::MultivariateDistribution, Q::MultivariateDistribution; kwargs...)
+    expectation(P, safe_logdiff(P, Q); kwargs...)
```

Comment:
What I wanted to suggest is to have only a single fallback definition. This should cover all cases:
```suggestion
function kldivergence(P::Distribution{V}, Q::Distribution{V}; kwargs...) where {V<:VariateForm}
    function logdiff(x)
        logp = logpdf(P, x)
        return (logp > oftype(logp, -Inf)) * (logp - logpdf(Q, x))
    end
    expectation(P, logdiff; kwargs..)
end
```